src/utils/storageFacade.utils.ts

```diff
@@ -1,122 +1,355 @@
+    for (let pkg of data) {
+      packages.push(dbPkgToPkg(pkg));
+    }
+  }
+
+  return packages;
 }
 
-export async function readAllPackages(sessionid:number){
-  let packageresult = await supabase //common syntax on JS: const {data,error} = await...
-    .from('packages') 
+/**
+ * Read a package by package reference string
+ * @param packageRef unique string ID retrieved from user-uploaded software bill of materials
+ * @param sessionId associated with one user
+ * @returns package if found in DB, null otherwise
+ */
+export async function readPackage(packageRef: string, sessionId: number) {
+  let { data, error } = await supabase
+    .from('packages')
     .select('*') //values are outputted first in, last out
-    .eq('sessionid',sessionid)
-  return packageresult;
-}
-/****************** READ VULNERABILITY **********************/
- export async function readVulnByPkg(packageid:number){
-  const {data}  = await supabase
-  .from ('junction')
-  .select('packageid,vulnerabilities(*)')
-  .eq('packageid',packageid)
-  return data;
- }
-
-
-//Task #3: create function to process convert CVEID string to cveid number - phase 2
-  export async function readVulnBySession(sessionid:number){
-    let {data} = await supabase
-    .from ('packages')
-    .select('junction!inner(packageid,vulnerabilities!inner(*))')
-    .eq('sessionid', sessionid)
-    return data;
+    .eq('sessionid', sessionId)
+    .eq('package_ref', packageRef);
+
+  if (error) {
+    logger.error(error.message);
+    return null;
+  } else if (data) {
+    if (data.length == 0) {
+      //No entry found
+      logger.error('Storage Facade: Package "' + packageRef + '" not found');
+      return null;
+    } else if (data.length > 1) {
+      //Duplicates (how should we handle?)
+      logger.error(
+        'Storage Facade: Duplicate packages found for "' +
+          packageRef +
+          '" in session ' +
+          sessionId
+      );
+      return dbPkgToPkg(data[0]);
+    } else {
+      return dbPkgToPkg(data[0]);
+    }
+  }
+}
+
+/**
+ * Write Package, updates if entry exists in database, inserts otherwise
+ * @param pkg to write to database
+ * @param sessionId associated with one user
+ * @returns supabase write status
+ */
+export async function writePackage(pkg: Package, sessionId: number) {
+  let { data, error, status } = await supabase
+    .from('packages')
+    .select('*')
+    .eq('sessionid', sessionId)
+    .eq('package_ref', pkg.ref);
+
+  if (error) {
+    logger.error(error.message);
+  } else if (data) {
+    if (data.length == 0) {
+      //INSERT
+      return insertPackage(pkg, sessionId);
+    } else {
+      //UPDATE
+      return updatePackage(pkg, sessionId);
+    }
+  }
+  return status;
+}
+
+/**
+ * Read all vulnerabilities in one session
+ * @param sessionId associated with one user
+ * @returns list of vulnerabilities, empty list if none matching present in DB
+ */
+export async function readVulnsBySession(sessionId: number) {
```

Comment:
same problem as read vulnby package - this function assumes sessionid is in junction table when it shouldn't be! must be rewritten entirely - refer to last function

Reply:
this is a nested junction, it joins vulnerabilities, junction, and packages
to ensure that you can read/write by sessionId (linked to vulnerabiliites by which package they're in)
You can test it in postman - I have a collection set up